### PR TITLE
use unique job id in notifications worker

### LINKED ISSUE #637: Certificate expiration warning results in 4 emails
Create a certificate on staging.  Wait for a few hours, and you'll get 4 notification emails saying that the certificate is expiring (staging certs only last 1 week).

We should update the notification logic so that it doesn't spam users.  Are we not updating the database to indicate when we last told them about expiry?

cc @SerpentBytes

### PR DESCRIPTION
**Pull request by SerpentBytes**
## Description

Closes #637. Uses unique job id in Notifications worker and properly updates the `lastNotified` field in the database. 

Steps to test

### Run the project
Notice the log messages
Create a new certificate (if needed)
Navigate to http://localhost:8025/ after 5 mins

### CODE REVIEW

**Comment by humphd** [comment]
This may also help with this error I see on staging:

```
{"code":"EMESSAGE","command":"DATA","level":"error","message":"Unable to send notification Message failed: 432 4.3.2 Concurrent connections limit exceeded. Visit https://aka.ms/concurrent_sending for more information. [Hostname=PH0PR07MB9202.namprd07.prod.outlook.com]","response":"432 4.3.2 Concurrent connections limit exceeded. Visit https://aka.ms/concurrent_sending for more information. [Hostname=PH0PR07MB9202.namprd07.prod.outlook.com]","responseCode":432,"stack":"Error: Message failed: 432 4.3.2 Concurrent connections limit exceeded. Visit https://aka.ms/concurrent_sending for more information. [Hostname=PH0PR07MB9202.namprd07.prod.outlook.com]\n    at SMTPConnection._formatError (/app/node_modules/nodemailer/lib/smtp-connection/index.js:790:19)\n    at SMTPConnection._actionSMTPStream (/app/node_modules/nodemailer/lib/smtp-connection/index.js:1690:34)\n    at SMTPConnection.<anonymous> (/app/node_modules/nodemailer/lib/smtp-connection/index.js:1158:22)\n    at SMTPConnection._processResponse (/app/node_modules/nodemailer/lib/smtp-connection/index.js:953:20)\n    at SMTPConnection._onData (/app/node_modules/nodemailer/lib/smtp-connection/index.js:755:14)\n    at TLSSocket.SMTPConnection._onSocketData (/app/node_modules/nodemailer/lib/smtp-connection/index.js:193:44)\n    at TLSSocket.emit (node:events:513:28)\n    at addChunk (node:internal/streams/readable:324:12)\n    at readableAddChunk (node:internal/streams/readable:297:9)\n    at TLSSocket.Readable.push (node:internal/streams/readable:234:10)","timestamp":"2023-04-18T00:47:41.438Z"}
```

**Review by humphd — CHANGES_REQUESTED** [review]

File: services/expiration/expiration-notification.server.ts [hunk 1]
```diff
@@ -189,8 +189,8 @@ const getExpiringDnsRecords = () => {
 };
 // function to add jobs
 async function addExpirationNotifications(type: string) {
-  let jobName = `${expirationNotificationQueueName}-${type}-expiry`;
-  return expirationNotificationQueue.add(jobName, {
+  let jobId = `${expirationNotificationQueueName}-${type}-expiry`;
+  return expirationNotificationQueue.add(expirationNotificationQueueName, jobId, {
```
Comment (humphd) [inline services/expiration/expiration-notification.server.ts:193]: I think the `jobId` has to be part of `data` passed in here.  Should you not wrap this in `{...}`:

```ts
return expirationNotificationQueue.add(expirationNotificationQueueName, { jobId }, {
```

Confirm that I'm right, but I think you have to do this.

**Review by humphd — CHANGES_REQUESTED** [review]

File: services/expiration/expiration-notification.server.ts (hunk 1 above)
Comment (humphd) [reply services/expiration/expiration-notification.server.ts:193]: You need to fix this, which is why I'm not approving this.

**Review by humphd — CHANGES_REQUESTED** [review]

File: services/expiration/expiration-notification.server.ts (hunk 1 above)
Comment (SerpentBytes) [reply services/expiration/expiration-notification.server.ts:193]: I don't see duplicate emails, but I could add this.

**Review by humphd — APPROVED** [review]
Let's try this. My inbox is filling up with staging certificate expiry warnings, so it would be nice to fix!

**Comment by SerpentBytes** [comment]
> Let's try this. My inbox is filling up with staging certificate expiry warnings, so it would be nice to fix!

On the same boat here.

**Review by Eakam1007 — COMMENTED** [review]
I am getting two emails for the expiry, do I need to change the setup to see this change? 
(Ignore the failure, mysql was not ready)
image

**Comment by SerpentBytes** [comment]
> I am getting two emails for the expiry, do I need to change the setup to see this change? 
> 
> (Ignore the failure, mysql was not ready)
> 
> image
> 
> 
I think you need to restart Redis container.

**Comment by humphd** [comment]
Restarting redis won't be enough, you'll also need to delete `redis-data/` and start it again

**Comment by humphd** [comment]
I also wonder if we have another bug on dev/staging, where we are sending more emails because we're within the 1 month threshold.  We should make it so that we don't send another email after we've sent one, period.

**Comment by SerpentBytes** [comment]
> I also wonder if we have another bug on dev/staging, where we are sending more emails because we're within the 1 month threshold. We should make it so that we don't send another email after we've sent one, period.

I could look into it if we don't need it by tomorrow.

**Comment by humphd** [comment]
Tomorrow is not the final deadline for fixes to this code.  We'll keep working on it for the foreseeable future, so no pressure.

**Review by Eakam1007 — APPROVED** [review]
Worked after delete `redis-data/`. I only got one email after several minuts. Thanks for the clarification

**Pull request merged**